### PR TITLE
networks/list.go updates for --format

### PR DESCRIPTION
**Pull request by ParkerVR**
Podman Network ls --format changes

https://docs.google.com/document/d/1YwNkk7t9o-MUbIDcvSMFXkWlnKR-OlGjwD_L5axKx50

Has an issue with 'make validate':

cmd/podman/networks/list.go:97:3: ineffectual assignment to `format` (ineffassign)
		format = report.NormalizeFormat(networkListOptions.Format)
		^
cmd/podman/networks/list.go:99:12: ineffectual assignment to `format` (ineffassign)
		headers, format = createListOut()

However I was following another file (https://github.com/containers/podman/pull/7973/commits/e9b667bb5f9c72fd9903bfa5efd081ad24a2482b podman containers ps) which does not receive this problem
Also - this does not use networkListOptions in createListOut, does it need to?

Signed-off-by: Parker Van Roy <pvanroy@redhat.com>

### CODE REVIEW

File: cmd/podman/networks/list.go [hunk 1]
```diff
@@ -88,19 +90,36 @@ func networkList(cmd *cobra.Command, args []string) error {
 		row += "\n"
 	}
 
-	format := "{{range . }}" + row + "{{end}}"
-	if !cmd.Flag("format").Changed {
-		format = headers + format
+	var headers, format string
+
+	if cmd.Flags().Changed("format") {
+		headers = ""
```
Comment (rhatdan) [inline cmd/podman/networks/list.go:96]: Not needed.  It defaults to this.

File: cmd/podman/networks/list.go [hunk 2]
```diff
@@ -88,19 +90,36 @@ func networkList(cmd *cobra.Command, args []string) error {
...
+	format = headers + "{{range . }}" + row + "{{end}}"
 	tmpl, err := template.New("listNetworks").Parse(format)
 	if err != nil {
 		return err
 	}
+
 	w := tabwriter.NewWriter(os.Stdout, 8, 2, 2, ' ', 0)
+	defer w.Flush()
+
 	if err := tmpl.Execute(w, nlprs); err != nil {
```
Comment (rhatdan) [inline cmd/podman/networks/list.go:110]: Just `return tmpl.Execute(w,nlprs)`

File: cmd/podman/networks/list.go [hunk 3]
```diff
@@ -88,19 +90,36 @@ func networkList(cmd *cobra.Command, args []string) error {
...
 		return err
 	}
-	return w.Flush()
+	return nil
+}
+
+func createListOut() (string, string) {
```
Comment (rhatdan) [inline cmd/podman/networks/list.go:116]: How about?

```
func createListOut() (string, string) {
	row := "{{.Name}}\\t{{.Version}}\\t{{.Plugins}}\\n"
	return report.NormalizeFormat(defaultHeaders), report.NormalizeFormat(row)
}
```

**Review by rhatdan — APPROVED** [review]

**Comment by ParkerVR** [comment]
wip, current code changes break :
➜  podman git:(format-networks/list) podman network ls
NAME    VERSION  PLUGINS
Error: template: listNetworks:2:2: executing "listNetworks" at <.Name>: can't evaluate field Name in type []network.ListPrintReports

File: cmd/podman/networks/list.go (hunk 3 above)
Comment (jwhonce) [reply cmd/podman/networks/list.go:104]: You could pull the headers into here as well.

**Comment by jwhonce** [comment]
@ParkerVR Check that  you have `{{range .}}...{end}} ` around the template where .Name is not found

**Comment by ParkerVR** [comment]
> @ParkerVR Check that you have `{{range .}}...{end}} ` around the template where .Name is not found

thanks for the advice, I think this should be properly fixed. I don't have any networks to test with, can you give instructions on how to generate some examples?

File: cmd/podman/networks/list.go [hunk 4]
```diff
@@ -8,7 +8,9 @@ import (
 	"text/tabwriter"
 	"text/template"
 
+	"github.com/containers/podman/v2/cmd/podman/parse"
 	"github.com/containers/podman/v2/cmd/podman/registry"
+	"github.com/containers/podman/v2/cmd/podman/report"
```
Comment (ashley-cui) [inline cmd/podman/networks/list.go:13]: might be part of it, the report pkg was move into containers/common

try rebasing, running a `make vendor-in-container` and see if these pop up/use these imports instead:
```
"github.com/containers/common/pkg/report"
"github.com/containers/podman/v2/cmd/podman/common"
```

**Comment by rhatdan** [comment]
Does the new containers/common with format fixes fix this PR?

**Comment by ashley-cui** [comment]
@rhatdan  they should work together, ie this pr should use the containers/common format fixes

File: cmd/podman/networks/list.go (hunk 4 above)
Comment (jwhonce) [reply cmd/podman/networks/list.go:13]: @ashley-cui @ParkerVR the PR I am working  on now will remove github.com/containers/podman/v2/cmd/podman/report

**Review by ashley-cui — COMMENTED** [review]
LGTM given green tests :)

**Comment by TomSweeneyRedHat** [comment]
and please remove the `[WIP]` in the title before merging.

**Comment by rhatdan** [comment]
Also if you are rebasing could you squash commits and change dwalsh->rhatdan or don't mention me at all.

**Comment by rhatdan** [comment]
@ParkerVR What is the scoop on this one?

**Comment by rhatdan** [comment]
@ParkerVR Any update with this?

**Comment by vrothberg** [comment]
Ping

File: cmd/podman/networks/list.go [hunk 5]
```diff
@@ -78,13 +79,37 @@ func networkList(cmd *cobra.Command, args []string) error {
...
 	switch {
-	case report.IsJSON(networkListOptions.Format):
-		return jsonOut(responses)
+	// quiet means we only print the network names
 	case networkListOptions.Quiet:
-		// quiet means we only print the network names
 		return quietOut(responses)
+
+	// TODO remove references to parse package
+	case parse.MatchesJSONFormat(networkListOptions.Format):
```
Comment (jwhonce) [inline cmd/podman/networks/list.go:87]: See https://github.com/containers/common/blob/master/pkg/report/validate.go

File: cmd/podman/networks/list.go [hunk 6]
```diff
@@ -78,13 +79,37 @@ func networkList(cmd *cobra.Command, args []string) error {
...
+		return jsonOut(responses)
+
+	// table or other format output
+	default:
+		return templateOut(responses, cmd)
+	}
+}
+
+func quietOut(responses []*entities.NetworkListReport) error {
```
Comment (jwhonce) [inline cmd/podman/networks/list.go:96]:
```suggestion
func quietOut(responses []*entities.NetworkListReport) {
```
I think this is cleaner if quietOut() is not following an interface.

File: cmd/podman/networks/list.go [hunk 7]
```diff
@@ -78,13 +79,37 @@ func networkList(cmd *cobra.Command, args []string) error {
...
+	for _, r := range responses {
+		fmt.Println(r.Name)
+	}
+	return nil
+}
+
+func jsonOut(responses []*entities.NetworkListReport) error {
+	prettyJSON, err := json.MarshalIndent(responses, "", "  ")
```
Comment (jwhonce) [inline cmd/podman/networks/list.go:104]: You should not import encoding/json. Add the following to network.go and just use the package global.

```
	// Pull in configured json library
	json = registry.JSONLibrary()
```

File: cmd/podman/networks/list.go [hunk 8]
```diff
@@ -99,13 +124,16 @@ func networkList(cmd *cobra.Command, args []string) error {
...
-	row := "{{.Name}}\t{{.Version}}\t{{.Plugins}}\n"
+
+	renderHeaders := strings.HasPrefix(networkListOptions.Format, "table ")
+	var row, format string
 	if cmd.Flags().Changed("format") {
-		renderHeaders = parse.HasTable(networkListOptions.Format)
 		row = report.NormalizeFormat(networkListOptions.Format)
+	} else { // 'podman network ls' equivalent to 'podman network ls --format="table {{ .Name }} {{ .Version }} {{ .Plugins }}" '
+		renderHeaders = true
+		row = defaultListRow
```
Comment (jwhonce) [inline cmd/podman/networks/list.go:134]: I would inline defaultListRow here rather than use a global.

File: cmd/podman/networks/list.go [hunk 9]
```diff
@@ -99,13 +124,16 @@ func networkList(cmd *cobra.Command, args []string) error {
 		"Labels":     "labels",
 		"ID":         "network id",
 	})
-	renderHeaders := true
-	row := "{{.Name}}\t{{.Version}}\t{{.Plugins}}\n"
+
+	renderHeaders := strings.HasPrefix(networkListOptions.Format, "table ")
```
Comment (jwhonce) [inline cmd/podman/networks/list.go:128]: See https://github.com/containers/common/blob/master/pkg/report/template.go#L171

File: cmd/podman/networks/list.go [hunk 10]
```diff
@@ -99,13 +124,16 @@ func networkList(cmd *cobra.Command, args []string) error {
...
+	var row, format string
 	if cmd.Flags().Changed("format") {
-		renderHeaders = parse.HasTable(networkListOptions.Format)
 		row = report.NormalizeFormat(networkListOptions.Format)
+	} else { // 'podman network ls' equivalent to 'podman network ls --format="table {{ .Name }} {{ .Version }} {{ .Plugins }}" '
+		renderHeaders = true
+		row = defaultListRow
 	}
-	format := parse.EnforceRange(row)
+	format = "{{range .}}" + row + "{{end}}"
```
Comment (jwhonce) [inline cmd/podman/networks/list.go:136]:
```suggestion
	format = report.EnforceRange(row)
```

File: cmd/podman/networks/list.go [hunk 11]
```diff
@@ -38,6 +38,7 @@ var (
 	networkListOptions entities.NetworkListOptions
 	filters            []string
 	noTrunc            bool
+	defaultListRow     = "{{.Name}}\t{{.Version}}\t{{.Plugins}}\n"
```
Comment (jwhonce) [inline cmd/podman/networks/list.go:41]:
```suggestion
	defaultListRow     = "{{.Id}}\t{{.Name}}\t{{.Version}}\t{{.Plugins}}\n"
```

This is the docker output today...

```shell
$ docker network ls
NETWORK ID          NAME                DRIVER          SCOPE
7fca4eb8c647        bridge              bridge          local
```

**Review by jwhonce — CHANGES_REQUESTED** [review]

**Comment by rhatdan** [comment]
@ParkerVR Could you rebase this PR?

**Comment by ParkerVR** [comment]
made a mistake rebasing friday that added some extraneous files, will get it rebased after I get out of class today

File: cmd/podman/networks/list.go (hunk 11 above)
Comment (ParkerVR) [reply cmd/podman/networks/list.go]: added, tried .Id but it's set for only .ID

**Review by jwhonce — APPROVED** [review]
LGTM

@mheon Are you good with the bump in common version?

**Comment by mheon** [comment]
Against master, to an already-tagged version? I see no problem.

/lgtm

**Pull request merged**